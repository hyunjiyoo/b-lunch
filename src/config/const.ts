import { UserType } from 'types';

const UPLOAD_PRESET = process.env.REACT_APP_CLOUDINARY_UPLOAD_PRESET as string;
const CLOUDINARY_CLOUD_NAME = process.env.REACT_APP_CLOUDINARY_CLOUD_NAME;
const CLOUDINARY_URL = `https://api.cloudinary.com/v2/${CLOUDINARY_CLOUD_NAME}/upload/q_50`;
const DEFAULT_IMAGE_URL = 'https://via.placeholder.com/400';

const initialUser: UserType = {
  uid: '',
  displayName: '',
  email: '',
  isAdmin: false,
  photoURL: '',
};

const MESSAGE = {
  LOGIN_INFO: '로그인을 먼저 해주세요.',
  ADMIN_INFO: '권한이 없습니다.',
  OPTION_INFO: '옵션을 먼저 선택해주세요.',
  CART_INFO: '장바구니에 상품이 추가되었습니다.\n장바구니로 이동하시겠습니까?',
};

const DELIVERY_COST = 3000;
const FREE_DELIVERTY_PRICE = 100000;
const LIMIT_PRODUCT_PRICE = 1000000;

export { UPLOAD_PRESET, CLOUDINARY_URL, DEFAULT_IMAGE_URL, initialUser, MESSAGE, DELIVERY_COST, FREE_DELIVERTY_PRICE, LIMIT_PRODUCT_PRICE };
