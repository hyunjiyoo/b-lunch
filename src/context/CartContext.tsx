import { createContext, useContext, useState } from 'react';

interface CartContextProps {
  count: number;
  updateCount: () => void;
}

const CartContext = createContext<CartContextProps>({} as CartContextProps);

interface CartProviderProps {
  children: JSX.Element;
}

export function CartProvider({ children }: CartProviderProps) {
  const [count, setCount] = useState<number>(0);

  const updateCount = () => {
    const cart = JSON.parse(localStorage.cart ?? JSON.stringify({}));
    setCount(Object.keys(cart).length);
  }

  return <CartContext.Provider value={{ count, updateCount }}>{children}</CartContext.Provider>;
}

export const useCart = () => useContext(CartContext);
