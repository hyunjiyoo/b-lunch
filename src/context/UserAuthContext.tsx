import { createContext, useContext, useEffect, useState } from 'react';
import { getUserInfo } from 'util/\bcommon';

interface UserAuthContextProps {
  isAdmin: boolean;
}

export const UserAuthContext = createContext<UserAuthContextProps>({ isAdmin: false });

interface UserAuthProps {
  children: JSX.Element;
}

export function UserAuthProvider({ children }: UserAuthProps) {
  const [isAdmin, setIsAdmin] = useState<boolean>(() => getUserInfo());

  useEffect(() => {
    setIsAdmin(() => getUserInfo());
  }, []);

  return <UserAuthContext.Provider value={{ isAdmin }}>{children}</UserAuthContext.Provider>;
}

export const useUserAuth = () => useContext(UserAuthContext);
