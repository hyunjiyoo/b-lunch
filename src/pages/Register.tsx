import { CLOUDINARY_URL, DEFAULT_IMAGE_URL, MESSAGE, UPLOAD_PRESET } from 'config/const';
import { addProduct } from 'db/database';
import { yupResolver } from '@hookform/resolvers/yup';
import * as yup from 'yup';
import { ChangeEvent, useEffect, useState } from 'react';
import { useForm } from 'react-hook-form';
import { ProductType } from 'types';
import * as R from 'ramda';
import { v4 as uuid } from 'uuid';
import Error from 'components/Error/Error';
import { getUserInfo } from 'util/\bcommon';
import { useNavigate } from 'react-router-dom';

const schema = yup.object({
  file: yup.mixed().test('filesize', '파일을 등록해주세요', (file: FileList) => file.length > 0),
  name: yup.string().trim().required('제품명을 입력해주세요.').min(2, '2자 이상 입력해주세요!'),
  price: yup
    .number()
    .transform((value) => (isNaN(value) ? undefined : value))
    .required('가격을 입력해주세요.')
    .integer('가격은 정수만 입력가능합니다.')
    .min(100, '100원 이상의 상품만 등록가능합니다.')
    .max(1000000, '1,000,000원 이하의 상품을 등록해주세요.'),
  category: yup.string().trim().required('카테고리를 입력해주세요.'),
  description: yup.string().trim().required('제품 설명을 입력해주세요.'),
  option: yup.string().trim().required('옵션을 입력해주세요.'),
});

export default function Register() {
  const navigate = useNavigate();
  const [previewImg, setPreviewImage] = useState<string>(DEFAULT_IMAGE_URL);
  const {
    register,
    handleSubmit,
    reset,
    formState: { errors },
  } = useForm<ProductType>({
    resolver: yupResolver(schema),
  });

  const previewImage = (event: ChangeEvent<HTMLInputElement>) => {
    const file = event.target.files as FileList;

    if (file.length > 0) {
      const src = URL.createObjectURL(file[0]);
      setPreviewImage(src);
    }
  };

  const imageUploadToCloudinary = async (file: File) => {
    const body = new FormData();
    body.append('file', file);
    body.append('upload_preset', UPLOAD_PRESET);

    return fetch(CLOUDINARY_URL, { method: 'POST', body })
      .then((res) => res.json())
      .then((data) => data.url);
  };

  const registerProduct = async (product: ProductType) => {
    try {
      const option = product.option.split(',').filter((opt: string) => opt.trim()).join();
      const { name, price, category, description, file } = product;
      const fileObj = (file as unknown as FileList)[0];
      const imageUrl = await imageUploadToCloudinary(fileObj);
      const id = uuid();
      const newItem = { id, name, price, category, description, option, imageUrl };
      console.log(newItem);
      addProduct(newItem);
      alert('새로운 제품이 성공적으로 등록되었습니다.');
      setPreviewImage(DEFAULT_IMAGE_URL);
      reset();
    } catch (error) {
      console.error(error);
    }
  };

  useEffect(() => {
    if (R.isEmpty(getUserInfo().uid)) {
      alert(MESSAGE.LOGIN_INFO);
      navigate('/');
    }
    
    if (!getUserInfo().isAdmin) {
      alert(MESSAGE.ADMIN_INFO);
      navigate('/');
    }
  }, [navigate]);

  return (
    <>
      <h2 className='text-lg font-semibold mt-10'>제품 등록</h2>
      <hr className='mt-2 mb-4' />
      <div className='flex flex-col lg:flex-row'>
        <div className='lg:w-350px h-350px w-full bg-slate-100 mr-4'>
          <img src={previewImg} alt='preview_image' className='w-full h-full object-contain' />
        </div>
        <form className='flex flex-col flex-1 gap-2' onSubmit={handleSubmit(registerProduct)}>
          <Error error={errors.file?.message}>
            <input
              type='file'
              className='border-2 p-2 text-sm outline-orange-400 mt-2'
              accept='image/*'
              {...register('file', {
                onChange: previewImage,
              })}
            />
          </Error>
          <Error error={errors.name?.message}>
            <input type='text' placeholder='제품명' className='text-sm border-2 p-2 outline-orange-400' {...register('name')} />
          </Error>
          <Error error={errors.price?.message}>
            <input type='number' placeholder='가격' className='text-sm border-2 p-2 outline-orange-400' {...register('price')} />
          </Error>
          <Error error={errors.category?.message}>
            <input type='text' placeholder='카테고리' className='text-sm border-2 p-2 outline-orange-400' {...register('category')} />
          </Error>
          <Error error={errors.description?.message}>
            <input type='text' placeholder='제품 설명' className='text-sm border-2 p-2 outline-orange-400' {...register('description')} />
          </Error>
          <Error error={errors.option?.message}>
            <input type='text' placeholder='옵션들(콤마(,)로 구분)' className='text-sm border-2 p-2 outline-orange-400' {...register('option')} />
          </Error>
          <button type='submit' className='bg-orange-500 text-white py-2 rounded-xl mt-3'>
            제품 등록하기
          </button>
        </form>
      </div>
    </>
  );
}
