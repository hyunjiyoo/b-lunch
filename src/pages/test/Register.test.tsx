import { render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import Register from 'pages/Register';
import { withRouter } from 'tests/utils';
import { DEFAULT_IMAGE_URL, MESSAGE } from 'config/const';
import { UserAuthContext } from 'context/UserAuthContext';
import { Route } from 'react-router-dom';

beforeEach(() => {
  window.alert = jest.fn();
});

afterEach(() => {
  jest.restoreAllMocks();
});

describe('Register', () => {
  it('renders all required form elements', () => {
    render(withRouter(<Route path='/' element={<Register />} />));

    const image = screen.getByRole('img') as HTMLImageElement;
    const button = screen.getByRole('button', { name: '제품 등록하기' }) as HTMLButtonElement;

    expect(image.alt).toBe('preview_image');
    expect(image.src).toBe(DEFAULT_IMAGE_URL);
    expect(button).toBeInTheDocument();

    expect(screen.getByPlaceholderText('제품명')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('제품명')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('제품명')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('제품 설명')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('옵션들(콤마(,)로 구분)')).toBeInTheDocument();
  });

  it('displays an alert message when user is not admin', async () => {
    render(
      withRouter(
        <Route
          path='/'
          element={
            <UserAuthContext.Provider value={{ isAdmin: false }}>
              <Register />
            </UserAuthContext.Provider>
          }
        />
      )
    );

    await waitFor(() => expect(alert).toHaveBeenCalledTimes(1));
    expect(alert).toHaveBeenCalledWith(MESSAGE.ADMIN_INFO);
  });
});
