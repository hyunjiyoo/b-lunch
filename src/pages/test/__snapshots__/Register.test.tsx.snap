// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register renders correctly 1`] = `
Array [
  <h2
    className="text-lg font-semibold mt-10"
  >
    제품 등록
  </h2>,
  <hr
    className="mt-2 mb-4"
  />,
  <div
    className="flex flex-col lg:flex-row"
  >
    <div
      className="lg:w-350px h-350px w-full bg-slate-100 mr-4"
    >
      <img
        alt="preview_image"
        className="w-full h-full object-contain"
        src="https://via.placeholder.com/400"
      />
    </div>
    <form
      className="flex flex-col flex-1 gap-2"
      onSubmit={[Function]}
    >
      <input
        accept="image/*"
        className="border-2 p-2 text-sm outline-orange-400 mt-2"
        name="file"
        onBlur={[Function]}
        onChange={[Function]}
        type="file"
      />
      <input
        className="text-sm border-2 p-2 outline-orange-400"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="제품명"
        type="text"
      />
      <input
        className="text-sm border-2 p-2 outline-orange-400"
        name="price"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="가격"
        type="number"
      />
      <input
        className="text-sm border-2 p-2 outline-orange-400"
        name="category"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="카테고리"
        type="text"
      />
      <input
        className="text-sm border-2 p-2 outline-orange-400"
        name="description"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="제품 설명"
        type="text"
      />
      <input
        className="text-sm border-2 p-2 outline-orange-400"
        name="option"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="옵션들(콤마(,)로 구분)"
        type="text"
      />
      <button
        className="bg-orange-500 text-white py-2 rounded-xl mt-3"
        type="submit"
      >
        제품 등록하기
      </button>
    </form>
  </div>,
]
`;
